engines/sword1/screen.cpp

```diff
@@ -168,23 +186,92 @@ void Screen::fnSetPalette(uint8 start, uint16 length, uint32 id, bool fadeUp) {
 		_targetPalette[(start + cnt) * 3 + 2] = palData[cnt * 3 + 2] << 2;
 	}
 	_resMan->resClose(id);
-	_isBlack = false;
-	if (fadeUp) {
-		_fadingStep = 1;
-		_fadingDirection = FADE_UP;
-		memset(_currentPalette, 0, 256 * 3);
-		_system->getPaletteManager()->setPalette(_currentPalette, 0, 256);
-	} else
-		_system->getPaletteManager()->setPalette(_targetPalette + 3 * start, start, length);
+
+	_system->getPaletteManager()->setPalette(_targetPalette + 3 * start, start, length);
+}
+
+void Screen::fnSetFadeTargetPalette(uint8 start, uint16 length, uint32 id, bool toBlack) {
+	byte *rgbData = nullptr;
+
+	if (toBlack) {
+		rgbData = const_cast<byte *>(_black);
+	} else {
+		rgbData = (byte *) _resMan->openFetchRes(id);
+	}
+
+	if (SwordEngine::_systemVars.wantFade) {
+		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
+		memcpy(_paletteFadeInfo.dstPalette + (start * 3), rgbData, length * 3);
+		memset(_paletteFadeInfo.srcPalette + (start * 3), 0, length * 3);
+	} else {
+		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
+	}
+
+	_resMan->resClose(id);
 }
 
-void Screen::fullRefresh() {
+void Screen::fullRefresh(bool soft) {
 	_fullRefresh = true;
-	_system->getPaletteManager()->setPalette(_targetPalette, 0, 256);
+	if (!soft)
+		_system->getPaletteManager()->setPalette(_targetPalette, 0, 256);
 }
 
-bool Screen::stillFading() {
-	return (_fadingStep != 0);
+int16 Screen::stillFading() {
+	if (SwordEngine::_systemVars.wantFade)
+		return _paletteFadeInfo.paletteStatus;
+	else
+		return 0;
+}
+
+
+void Screen::fadePalette() {
+	_paletteFadeInfo.fadeCount++;
+
+	if (_paletteFadeInfo.fadeCount == _paletteFadeInfo.paletteIndex) {
+		_paletteFadeInfo.fadeCount = 0;
+
+		if (_paletteFadeInfo.paletteStatus == NO_FADE)
+			return;
+
+		// Set the whole palette to the current source.
+		// But first, remember that these are 6 bit values, and
+		// they have to be shifted left two times before using them.
+		byte outPal[256 * 3];
+		for (int i = 0; i < 256 * 3; i++) {
+			// Remember that we previously allowed the RGB values
+			// to go below zero? It's time to account for that;
+			// This contributes to producing the correct palette fading effect.
+			int8 curValueSigned = (int8)_paletteFadeInfo.srcPalette[i];
+			if (curValueSigned < 0)
+				curValueSigned = 0;
+
+			outPal[i] = ((byte)curValueSigned) << 2;
+		}
+
+		_system->getPaletteManager()->setPalette((const byte *)outPal, 0, 256);
+
+		_paletteFadeInfo.paletteCount--;
+
+		if (_paletteFadeInfo.paletteCount == 0) {
+			_paletteFadeInfo.paletteStatus = NO_FADE;
+		} else {
+			if (_paletteFadeInfo.paletteStatus == FADE_DOWN) {
```

Comment:
Perhaps, I am too picky, but I'd write something like:

```C++
for (int i = 0; i < 256 * 3; i++) {
   if (_paletteFadeInfo.srcPalette[i] > 0 && (int8)_paletteFadeInfo.srcPalette[i] < (int8)_paletteFadeInfo.dstPalette[i])
    _paletteFadeInfo.srcPalette[i] += _paletteFadeInfo.paletteStatus;
 ```

Since `FADE_UP = 1` and `FADE_DOWN = -1`

Reply:
Your way of writing this is certainly more concise and shorter; unfortunately I have tested it and it seems to break fades 🙁 It appears the two conditions can't be used together. 

If you are okay with it I'd leave the verbose code on

Reply:
This indeed can't work as when fading up from black, srcPalette will be 0 and fading will never happen.
Maybe CLIP could be used but it seems safer to keep it the way it's written here as there could be some wrapping (if we fade from white to black for example).